src/layer/layer_impl.rs

```diff
@@ -13,11 +17,15 @@ use crate::{
 
 use super::*;
 
+static SPAN_DATA: LazyLock<RwLock<HashMap<tracing::span::Id, SpanData, BuildHasherDefault<FNV1aHasher64>>>> = LazyLock::new(|| {
+    RwLock::new(HashMap::with_hasher(BuildHasherDefault::<FNV1aHasher64>::default()))
+});
+
 struct SpanData {
     fields: Box<[FieldValueIndex]>,
     activity_id: [u8; 16], // // if set, byte 0 is 1 and 64-bit span ID in the lower 8 bytes
     related_activity_id: [u8; 16], // if set, byte 0 is 1 and 64-bit span ID in the lower 8 bytes
-    start_time: SystemTime,
+    start_time: RwLock<SystemTime>,
```

Comment:
The old code got away without locking this because it was being boxed into a dyn trait object, which hid the problem from the compiler. Now it sees that there is a data race that requires a lock.

At a higher level though, there is no data race, and I wish this extra locking could be avoided.

Reply:
I don't believe you. This isn't needed.

Reply:
😥

SpanData can only be Sync if all its members are Sync. Mutable fields are not Sync unless wrapped by something that implements Sync, which of the interior mutability types is RwLock or Mutex. If I manually tag the struct as Sync without ensuring no reads can happen at the same time as the write, it would technically be UB.

In theory, it is possible to start a struct on one thread and end it on another. Both operations only require a const reference from the logger's point of view - if a layer wants to implement any sort of ordering or mutation within it's implementation, it must do so safely, which is what the code is trying to.

The outer RwLock doesn't protect against interior mutability at all. A mutex would, but that's too aggressive for what we need, which is to ensure enter and exit don't encounter any data races when writing/reading this value.

Reply:
The outer lock protects you just fine because it ensures you can only access the reference to the interior struct while the outer lock is held.

Reply:
Multiple threads can hold a Read lock and still mutate an interior field though. That's why RwLock explicitly doesn't implement Sync unless each field does, while Mutex implements Sync always. Statics must be Sync.

Reply:
That is to say:
Statics must be Sync. Which means the RwLock must be Sync.
The RwLock is Sync if the thing it contains is Sync.
The struct is Sync only if its fields are all Sync.
A mutable field is not Sync unless it guarantees it cannot be mutated when any other reference to it exists, otherwise it is UB.
The mutable field must therefore be locked or atomically changeable.
Atomic<SystemTime> or RwLock<SystemTime>, take your pick. RwLock is conceptually nicer, especially since SystemTime is not guaranteed to be any particular size or underlying type.

Reply:
I think I see what you're saying. Here's my take.

Statics do not need to be Sync. The restriction is "The type must have the Sync trait bound to allow thread-safe access", but that only means "if you want safe code to be guaranteed by the compiler to have no race conditions, the static must be Sync".

The issue with RwLock is that there can be multiple threads that all see a const ref to T, so if it's bad for multiple threads to all read the same instance of T at the same time (e.g. Cell) then RwLock won't provide proper protection.

The docs indicate that SystemTime has an auto impl of Sync and Send, so it should all be fine.

If it weren't fine then the second RwLock wouldn't fix it. To make a non-Sync variable thread-safe you need a Mutex.

Interior mutability only comes into play if the mutation happens while the outer lock is not held. That's not happening here -- you only mutate data.start_time while you hold the lock on span_data.